src/EditorFeatures/Core/Implementation/CommentSelection/CommentUncommentSelectionCommandHandler.cs

```diff
@@ -83,115 +63,53 @@ public bool ExecuteCommand(UncommentSelectionCommandArgs args, CommandExecutionC
-            var service = document.GetLanguageService<ICommentSelectionService>();
-            if (service != null)
-            {
-                return service;
-            }
-
-            return null;
+            return operation == Operation.Comment
+                ? EditorFeaturesResources.Comment_Selection
+                : EditorFeaturesResources.Uncomment_Selection;
         }
 
-        private void Format(ICommentSelectionService service, ITextSnapshot snapshot, IEnumerable<ITrackingSpan> changes, CancellationToken cancellationToken)
+        protected override string GetMessage(Operation operation)
         {
-            var document = snapshot.GetOpenDocumentInCurrentContextWithChanges();
-            if (document == null)
-            {
-                return;
-            }
-
-            var textSpans = changes.Select(s => s.GetSpan(snapshot).Span.ToTextSpan()).ToImmutableArray();
-            var newDocument = service.FormatAsync(document, textSpans, cancellationToken).WaitAndGetResult(cancellationToken);
-            newDocument.Project.Solution.Workspace.ApplyDocumentChanges(newDocument, cancellationToken);
+            return operation == Operation.Comment
+                ? EditorFeaturesResources.Commenting_currently_selected_text
+                : EditorFeaturesResources.Uncommenting_currently_selected_text;
         }
 
-        internal enum Operation { Comment, Uncomment }
-
         /// <summary>
         /// Add the necessary edits to the given spans. Also collect tracking spans over each span.
         ///
         /// Internal so that it can be called by unit tests.
         /// </summary>
-        internal void CollectEdits(
+        internal override Task<CommentSelectionResult> CollectEdits(
             Document document, ICommentSelectionService service, NormalizedSnapshotSpanCollection selectedSpans,
-            List<TextChange> textChanges, List<ITrackingSpan> trackingSpans, Operation operation, CancellationToken cancellationToken)
+            Operation operation, CancellationToken cancellationToken)
         {
+            var spanTrackingList = new List<CommentTrackingSpan>();
+            var textChanges = new List<TextChange>();
             foreach (var span in selectedSpans)
             {
                 if (operation == Operation.Comment)
                 {
-                    CommentSpan(document, service, span, textChanges, trackingSpans, cancellationToken);
+                    CommentSpan(document, service, span, textChanges, spanTrackingList, cancellationToken);
                 }
                 else
                 {
-                    UncommentSpan(document, service, span, textChanges, trackingSpans, cancellationToken);
+                    UncommentSpan(document, service, span, textChanges, spanTrackingList, cancellationToken);
                 }
             }
+            return Task.FromResult(new CommentSelectionResult(textChanges, spanTrackingList, operation));
```

Comment:
1. newline after `}`
2. are any impls of this actually async?  if not, make this synchronous.

Reply:
Yes, the toggle one is async since it's accessing other services (and in c# the syntax tree)